gbmi/utils/hash.py

```diff
@@ -0,0 +1,94 @@
+"""
+Generate stable hashes for Python data objects.
+Contains no business logic.
+
+The hashes should be stable across interpreter implementations and versions.
+
+Supports dataclass instances, datetimes, and JSON-serializable objects.
+
+Empty dataclass fields are ignored, to allow adding new fields without
+the hash changing. Empty means one of: None, '', (), [], or {}.
+
+The dataclass type is ignored: two instances of different types
+will have the same hash if they have the same attribute/value pairs.
+
+"""
+import dataclasses
+import datetime
+import hashlib
+import json
+from collections.abc import Collection
+from typing import Any
+from typing import Dict
+
+# Implemented for https://github.com/lemon24/reader/issues/179
```

Comment:
Did you write this code @euanong ?  Do we need to stick some license notice at the top?